src/libmongoc/src/mongoc/mongoc-write-command.c

```diff
@@ -581,13 +581,18 @@ _mongoc_write_opmsg (mongoc_write_command_t *command,
                &txn_number_iter,
                ++parts.assembled.session->server_session->txn_number);
          }
+         bson_error_t original_error = {0};
       retry:
          ret = mongoc_cluster_run_command_monitored (
             &client->cluster, &parts.assembled, &reply, error);
 
          if (parts.is_retryable_write) {
             _mongoc_write_error_handle_labels (
-               ret, error, &reply, server_stream->sd->max_wire_version);
+               ret,
+               error,
+               &reply,
+               server_stream->sd->max_wire_version,
+               &original_error);
```

Comment:
On the first operation, the `original_error` is zeroed. That may result in zeroing out `error`.

Additionally, the original reply will need to be saved.

Consider updating the logic to store the original error and reply when retrying. And restoring the original error and reply when a retry fails with a `NoWritesPerformed` label.

Here is a suggested change: https://github.com/kevinAlbs/mongo-c-driver/commit/5ea4a339b01048005a80f93ddb8d64b871d0a722